### PR TITLE
virttest/test_setup.py: check 1G hugepage support.

### PR DESCRIPTION
**Pull request by CongSmile**
Signed-off-by: Cong Li <coli@redhat.com>

id: 1204060

Previous PR under virt-test is:
autotest/virt-test#2277

### CODE REVIEW

File: virttest/test_setup.py [hunk 1]
```diff
@@ -316,6 +329,43 @@ def __init__(self, params):
...
+        The flag in cpuinfo should be "pdpe1gb".
+        """
+        error_context.context("Check whether the host support 1G hugepage")
+        cpuinfo = file('/proc/cpuinfo').read()
+        cpu_flag = "pdpe1gb"
+        if cpu_flag not in cpuinfo:
+            raise HPNotSupportedError("Host does not support 1G support,"
+                                      "no such %s flag in host." % cpu_flag)
+        self.check_1G_hugepage_kernel_cmdline()
+        return
```
Comment (PandaWei) [inline virttest/test_setup.py]: I wonder if 'return' is necessary?

Comment (CongSmile) [inline virttest/test_setup.py]: Yes, it's not a must, will update it.

Thanks.

**Comment by CongSmile** [comment]
Hi @PandaWei 

I have removed "return()" in the new add func().
Could you help review it again?

Thanks. :)

File: virttest/test_setup.py [hunk 2]
```diff
@@ -316,6 +329,40 @@ def __init__(self, params):
...
+        error_context.context("Check host kernel cmdline support 1G hugepage")
+        kernel_cmdline = file('/proc/cmdline').read()
+        if "hugepagesz=1G" not in kernel_cmdline:
+            raise HPSetupError("Please check whether the host kernel"
+                               "cmdline has hugepagesz=1G option\n"
+                               "Host kernel cmdline is %s" % kernel_cmdline)
+
+    def check_hugepage_size_as_expected(self):
+        """
+        Check whether the hugepage size as excepted
```
Comment (PandaWei) [inline virttest/test_setup.py]: typo!
excepted?

Comment (PandaWei) [inline virttest/test_setup.py]: May be you can check it by yourself before pushing by set spell of VIM.

File: virttest/test_setup.py [hunk 3]
```diff
@@ -316,6 +329,40 @@ def __init__(self, params):
...
+        """
+        Check whether host kernel cmdline support 1G hugepage support.
+        Needs option "hugepagesz=1G" in kernel cmdline.
+        """
+        error_context.context("Check host kernel cmdline support 1G hugepage")
+        kernel_cmdline = file('/proc/cmdline').read()
+        if "hugepagesz=1G" not in kernel_cmdline:
+            raise HPSetupError("Please check whether the host kernel"
+                               "cmdline has hugepagesz=1G option\n"
+                               "Host kernel cmdline is %s" % kernel_cmdline)
```
Comment (PandaWei) [inline virttest/test_setup.py]: HPNotSupportedError better than  HPSetupError.

File: virttest/test_setup.py [hunk 4]
```diff
@@ -64,6 +64,14 @@ class HPNotSupportedError(Exception):
...
 
 
+class HPSetupError(Exception):
+
+    """
+    Thrown when hugepages setup error.
+    """
+    pass
+
+
```
Comment (PandaWei) [inline virttest/test_setup.py]: In my opinion,
No need to declare new Error.
HPNotSupportedError is enough.

File: virttest/test_setup.py [hunk 5]
```diff
@@ -64,6 +64,14 @@ class HPNotSupportedError(Exception):
     pass
 
 
+class HPSetupError(Exception):
+
+    """
+    Thrown when hugepages setup error.
+    """
+    pass
```
Comment (will-Do) [inline virttest/test_setup.py]: As you raise this Exception with detail msg, it's better to implement '__str__()' for this class

File: virttest/test_setup.py [hunk 6]
```diff
@@ -316,6 +329,40 @@ def __init__(self, params):
...
+
+    def check_hugepage_size_as_expected(self):
+        """
+        Check whether the hugepage size as expected
+        """
+        error_context.context("Check whether the hugepage size as expected")
+        if self.hugepage_size != self.expected_hugepage_size:
+            raise HPSetupError("The current hugepage size on host does"
+                               "not match the expected hugepage size.\n")
+
```
Comment (PandaWei) [inline virttest/test_setup.py]: As same as above.

File: virttest/test_setup.py [hunk 7]
```diff
@@ -316,6 +329,40 @@ def __init__(self, params):
...
+        Check whether host kernel cmdline support 1G hugepage support.
+        Needs option "hugepagesz=1G" in kernel cmdline.
+        """
+        error_context.context("Check host kernel cmdline support 1G hugepage")
+        kernel_cmdline = file('/proc/cmdline').read()
+        if "hugepagesz=1G" not in kernel_cmdline:
+            raise HPSetupError("Please check whether the host kernel"
+                               "cmdline has hugepagesz=1G option\n"
+                               "Host kernel cmdline is %s" % kernel_cmdline)
+
```
Comment (will-Do) [inline virttest/test_setup.py]: Maybe you can implement these two functions as more common methods(not specified 1G hugepage, use param insead): check_hugepage_support and check_hugepage_kernel_cmdline, so in check_hugepage_support, you can check the cpu flag or just pass.

Comment (will-Do) [inline virttest/test_setup.py]: BTW, another way to implement `check_1G_hugepage_kernel_cmdline` is abstract it as 'check_kernel_cmdline(pattern)` into utils_misc.py, then check whatever pattern(such as 'hugepage=1G') you want.

**Comment by CongSmile** [comment]
Hi @will-Do and @PandaWei 

I have updated the PR, could you help review it again?

Thanks a lot.

**Comment by CongSmile** [comment]
@xutian 

Could you help review this PR?

Thanks.

File: virttest/test_setup.py (hunk 7 above)
Comment (PandaWei) [inline virttest/test_setup.py]: @CongSmile 
I think @will-Do put forward a good suggestion.
You can divide this patch and extract some new common function into utils_misc.py.
If you don't have time to do it . I'm OK for this PR.

Comment (CongSmile) [inline virttest/test_setup.py]: Fine, I will update it.

**Comment by CongSmile** [comment]
@PandaWei 

Updated. could you help check it again?

Thanks.

File: virttest/test_setup.py [hunk 8]
```diff
@@ -317,6 +323,31 @@ def __init__(self, params):
 
         self.target_hugepages = target_hugepages
 
+    def check_hugepage_support(self):
+        """
+        Check whether the host support hugepage.
+        Need check the cpu flag and kernel CML.
+        """
+        error_context.context("Check whether the host support hugepage.")
+        utils_misc.check_cpu_flag(self.hugepage_cpu_flag)
```
Comment (xutian) [inline virttest/test_setup.py]: self.hugepage_cpu_flag in utils_misc.get_cpu_flags()

File: virttest/utils_misc.py [hunk 9]
```diff
@@ -907,6 +907,35 @@ def parallel(targets):
...
+    params: cpu_flag: The cpu flag which is need checked on host.
+    """
+    error_context.context("Check whether the host support %s." % cpu_flag)
+    cpuinfo = file('/proc/cpuinfo').read()
+    if cpu_flag not in cpuinfo:
+        raise exceptions.TestNAError("Host does not support %s flag in host." %
+                                     cpu_flag)
+
+
+def check_kernel_cmdline(match_str):
```
Comment (xutian) [inline virttest/utils_misc.py]: def get_kernel_cmdline():
      return xxx

File: virttest/test_setup.py [hunk 10]
```diff
@@ -317,6 +323,31 @@ def __init__(self, params):
...
+        self.check_hugepage_kernel_cmdline()
+
+    def check_hugepage_kernel_cmdline(self):
+        """
+        Check whether host kernel cmdline support hugepage support.
+        """
+        error_context.context("Check host kernel cmdline support hugepage")
+        utils_misc.check_kernel_cmdline(self.hugepage_match_str)
```
Comment (xutian) [inline virttest/test_setup.py]: return (self.hugepage_match_str in utils_misc.get_kernel_cmdline()
            and  self.hugepage_cpu_flag in utils_misc.get_cpu_flags() )

**Comment by CongSmile** [comment]
@xutian 

Updated. 

Could you help review it again?

Thanks.

File: virttest/utils_misc.py [hunk 11]
```diff
@@ -3278,6 +3278,15 @@ def rm_ker_cmd(kernel_cmdline, kernel_param):
     return kernel_cmdline
 
 
+def get_ker_cmd():
+    """
+    Get the kernel cmdline.
+
```
Comment (xutian) [inline virttest/utils_misc.py]: import avocado.utils.genio 

returun genio.read_file('/proc/cmdline')

**Comment by CongSmile** [comment]
@xutian

Updated.

Could you help review it again?

Thanks.

**Comment by xutian** [comment]
looks fine.  merged thanks @CongSmile

**Pull request merged**